sdk/core/Azure.Core/src/Shared/OperationInternalBase.cs

```diff
@@ -198,43 +236,37 @@ private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToke
 
             try
             {
-                return await UpdateStateAsync(async, cancellationToken).ConfigureAwait(false);
-            }
-            catch (Exception e)
-            {
-                scope.Failed(e);
-                throw;
-            }
-        }
-
-        protected async ValueTask<Response> ApplyStateAsync(bool async, Response response, bool hasCompleted, bool hasSucceeded, RequestFailedException? requestFailedException, bool throwIfFailed = true)
-        {
-            RawResponse = response;
+                var state = await UpdateStateAsync(async, cancellationToken).ConfigureAwait(false);
+                RawResponse = state.RawResponse;
+                if (!state.HasCompleted)
+                {
+                    return state.RawResponse;
+                }
 
-            if (!hasCompleted)
-            {
-                return response;
-            }
+                if (!state.HasSucceeded && state.OperationFailedException == null)
+                {
+                    var exception = async
+                        ? await _diagnostics.CreateRequestFailedExceptionAsync(state.RawResponse).ConfigureAwait(false)
+                        : _diagnostics.CreateRequestFailedException(state.RawResponse);
+                    state = OperationState.Failure(state.RawResponse, exception);
+                }
 
-            HasCompleted = true;
```

Comment:
I would like to understand more why this race condition has to be fixed this way.  I don't like that we are forcing an extra call even when the operation is completed (unless there is some nuance in your statement there).

Reply:
I read the OP comment of `To mitigate this problem` we are adding an extra call, but it was referring to the previous bullet and how it was previously fixed.  My understanding is now that the extra call is being removed.

I still would like to understand if there is a simpler way to resolve the race condition and allow the occasional thread to need an extra call

Reply:
Depends on what do you mean by "simpler". If we make an extra call, then we need to throw away its response if operation is in a final state. Otherwise, `Response.FromValue(Value, rawResponse)` may use `Value` from one request and `rawResponse` from another.
There are two options how to achieve it: using locks/semaphores or using state object that can be read/replaced in atomic operation. Since we have `OperationInternal` and `OperationInternal<T>`, there will be at least two objects (one derived from another). I can implement this version in a separate branch so we can compare.